### PR TITLE
Base implementation of Bond concept

### PR DESCRIPTION
**Pull request by perwestling**
- Issuer is a class on the same level as Corporation, Company
- Bond is a class on the same level as Share
- An Issuer can have a number of Bonds
- Presented in stock view if title has any Issuer
- Presented in Entity view if title has any Issuer
- Bonds Bought/Sold following share rules, using its price
  but no limit on how many bonds a player can own or sell
- Pays revenue to player owner each OR
- Bonds and Issuers shown on Spreadsheet, if title has any Issuer
- Bonds (if any own) shown on player chart

This commit is followed by a commit that adds auto buy of bonds,
and then a commit that adds use of Bonds in 1893.

### CODE REVIEW

**Comment by perwestling** [comment]
Screen shots how it looks when used in a title can be seen in #5990. Now this commit should have no effect on existing titles as no title yet uses Issuers and Bonds.

File: assets/app/view/game/buy_sell_bonds.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        return h(:div, children) unless children.empty?
+
+        nil
+      end
+
+      def render_buy_market_bonds
+        return [] unless @step.current_actions.include?('buy_bonds')
+
+        @game.bundles_for_issuer(@issuer, @issuer).map do |bundle|
+          next unless @step.can_buy_bonds?(@current_entity, bundle)
```
Comment (tobymao) [inline assets/app/view/game/buy_sell_bonds.rb:34]: these can have nils, need to compact

File: assets/app/view/game/buy_sell_bonds.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+        children = []
+
+        children.concat(render_buy_market_bonds)
+        children.concat(render_sell_bonds)
+
+        children.compact!
+        return h(:div, children) unless children.empty?
+
+        nil
```
Comment (tobymao) [inline assets/app/view/game/buy_sell_bonds.rb:27]: i don't like the pattern of returning nils here

**Comment by tobymao** [comment]
why do we need this? what is a bond?

File: assets/app/view/game/issuer.rb [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+            on: { click: ->(event) { select_issuer(event) } },
+          }
+          if selected?
+            props[:style][:backgroundColor] = 'lightblue'
+            props[:style][:color] = 'black'
+            props[:style][:border] = '1px solid'
+          end
+          props[:style][:display] = @display
+
+          header_text = @game.respond_to?(:issuer_header) ? @game.issuer_header(@issuer) : 'ISSUER'
```
Comment (tobymao) [inline assets/app/view/game/issuer.rb:78]: why do you have respond to here?

File: assets/app/view/game/issuer.rb [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+          bond_props = {
+            style: {
+              paddingRight: '1.3rem',
+            },
+          }
+
+          pool_bonds = @issuer.bonds.select(&:owned_by_issuer?)
+          pool_rows = []
+          if pool_bonds.any?
```
Comment (tobymao) [inline assets/app/view/game/issuer.rb:116]: unless empty

File: assets/app/view/game/issuer.rb [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+
+      def toggle_desc(event, issuer)
+        event.JS.stopPropagation
+        elm = Native(@hidden_divs[issuer.sym]).elm
+        elm.style.display = elm.style.display == 'none' ? 'grid' : 'none'
+      end
+
+      def render_issuer_on_card(issuer)
+        extra = []
+        if @game.respond_to?(:issuer_revenue_str) && @game.issuer_status_str(issuer)
```
Comment (tobymao) [inline assets/app/view/game/issuer.rb:163]: do you need all these respond to?

File: assets/app/view/game/issuers.rb [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+module View
+  module Game
+    class Issuers < Snabberb::Component
+      needs :game
+      needs :owner, default: nil
+
+      def render
+        bonds = @game.all_issuers.flat_map do |i|
+          bonds_of_issuer = @owner.bonds_of(i)
+          next unless bonds_of_issuer.any?
```
Comment (tobymao) [inline assets/app/view/game/issuers.rb:12]: if empty

File: assets/app/view/game/round/stock.rb [hunk 7]
```diff
@@ -62,6 +64,7 @@ def render
           children << h(SpecialBuy) if @current_actions.include?('special_buy')
           children.concat(render_failed_merge) if @current_actions.include?('failed_merge')
           children.concat(render_corporations)
+          children.concat(render_issuers) unless @game.all_issuers.none?
```
Comment (tobymao) [inline assets/app/view/game/round/stock.rb:67]: empty?

File: assets/app/view/game/round/stock.rb [hunk 8]
```diff
@@ -265,6 +268,33 @@ def render_mergeable_entities
...
+            h(:div, props, children)
+          end.compact
+        end
+
+        def render_issuer_input(issuer)
+          inputs = [
+            h(BuySellBonds, issuer: issuer),
+          ]
+          inputs = inputs.compact
+          h('div.margined_bottom', { style: { width: '20rem' } }, inputs) if inputs.any?
```
Comment (tobymao) [inline assets/app/view/game/round/stock.rb:295]: unless empty

**Pull request closed without merging**